### PR TITLE
Move roletemplate out of cluster lifecycle

### PR DESCRIPTION
**Pull request by dramich**
Problem:
cluster-roletemplate-sync was cluster scoped which means with 2k clusters there will 2k annotations and 2k finalizers on each roleTemplate. This causes slowdown when listing and also makes the object impossible to look at in general. 

Solution:
Move cluster-roletemplate-sync to a management level lifecycle to handle the downstream cluster deletion of the cluster roles and setup a cluster scoped handler to handle syncing the roleTemplate to clusterTemplate on changes.

issue: https://github.com/rancher/rancher/issues/23923

### CODE REVIEW

File: pkg/controllers/management/auth/role_template_handler.go [hunk 1]
```diff
@@ -1,47 +1,105 @@
...
+			continue
+		}
+
+		err = userContext.RBAC.ClusterRoles("").Delete(b.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			// User context clusterRole doesn't exist
+			if !apierrors.IsNotFound(err) {
+				allErrors = append(allErrors, err)
+			}
+			continue
```
Comment (rmweir) [inline pkg/controllers/management/auth/role_template_handler.go]: nit: probably don't need continue

Comment (dramich) [reply pkg/controllers/management/auth/role_template_handler.go]: It's not technically necessary but it's there to be clear about intentions and if something is added below this it won't cause a bug. grb-sync follows a similar pattern.

**Comment by dramich** [comment]
> Should there be an upgrade test?

The tests that were updated cover upgrade

File: pkg/api/store/globalrolebindings/globalrolebinding_store.go [hunk 2]
```diff
@@ -9,7 +9,10 @@ import (
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-const GrbVersion = "field.cattle.io/grbUpgrade"
+const (
+	GrbVersion    = "field.cattle.io/grbUpgradeCluster"
```
Comment (dramich) [inline pkg/api/store/globalrolebindings/globalrolebinding_store.go]: Update to use `cleanup.cattle.io`

File: pkg/api/store/roletemplate/store.go [hunk 3]
```diff
@@ -10,7 +10,10 @@ import (
 	"k8s.io/apimachinery/pkg/labels"
 )
 
-const RTVersion = "field.cattle.io/rtUpgrade"
+const (
+	RTVersion    = "field.cattle.io/rtUpgradeCluster"
```
Comment (dramich) [inline pkg/api/store/roletemplate/store.go]: Update to use `cleanup.cattle.io`

File: pkg/controllers/management/auth/role_template_handler.go [hunk 4]
```diff
@@ -1,47 +1,105 @@
...
-	prtbIndexer cache.Indexer
-	prtbClient  v3.ProjectRoleTemplateBindingInterface
+	prtbIndexer    cache.Indexer
+	prtbClient     v3.ProjectRoleTemplateBindingInterface
+	clusters       v3.ClusterInterface
+	clusterManager *clustermanager.Manager
 }
 
-func newRoleTemplateHandler(management *config.ManagementContext) *roleTemplateHandler {
+func newRoleTemplateHandler(management *config.ManagementContext, clusterManager *clustermanager.Manager) v3.RoleTemplateLifecycle {
```
Comment (cjellick) [inline pkg/controllers/management/auth/role_template_handler.go]: Since this is now a lifecycle, can we name the function accordingly?

Comment (cjellick) [reply pkg/controllers/management/auth/role_template_handler.go]: file name as well

Comment (cjellick) [reply pkg/controllers/management/auth/role_template_handler.go]: and the name `roleTemplateHandler` the functions are attached to

**Comment by dramich** [comment]
Offline discussion: update the user controller sync to validate the clusterrole has a roletemplate in management and delete clusterrole if there isn't a corresponding roletemplate

File: pkg/controllers/user/rbac/handler_base.go [hunk 5]
```diff
@@ -199,7 +195,8 @@ func (m *manager) createClusterRole(rt *v3.RoleTemplate) error {
 	logrus.Infof("Creating clusterRole for roleTemplate %v (%v).", rt.DisplayName, rt.Name)
 	_, err := m.clusterRoles.Create(&rbacv1.ClusterRole{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: rt.Name,
+			Name:        rt.Name,
+			Annotations: map[string]string{clusterRoleOwner: rt.Name},
 		},
```
Comment (mrajashree) [inline pkg/controllers/user/rbac/handler_base.go]: v1.Roles are also created with the name of the roleTemplate, should we track them and delete when roleTemplate no longer exists by adding annotations for them too here? https://github.com/rancher/rancher/blob/master/pkg/controllers/management/auth/manager.go#L735

Comment (dramich) [reply pkg/controllers/user/rbac/handler_base.go]: Well, those roletemplates are getting created in management so they could just be an owner reference but I don't know if that is intended behavior and outside the scope of this PR.

File: pkg/controllers/management/auth/role_template_handler.go (hunk 4 above)
Comment (dramich) [reply pkg/controllers/management/auth/role_template_handler.go]: Renamed everything

**Pull request merged**